### PR TITLE
feat(dapi): tenovi integration

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport#812

### Dependencies

- Upstream [992](https://github.com/metriport/metriport-internal/pull/992)

### Description

- Tenovi is integrated
- getBodyData and getBiometricsData is supported (query 1 patient at a time)
- Webhooks included
- SDK updated
- Docs guide to Tenovi connection

### Screenshots
Screenshot 2023-08-23 at 10 18 37 AM
Screenshot 2023-08-23 at 10 19 55 AM
Screenshot 2023-08-24 at 2 00 18 PM
Screenshot 2023-08-24 at 2 00 43 PM

### Release Plan

- Test on staging
- Let the CX know 
- Nothing special

### CODE REVIEW

File: docs/devices-api/api-reference/get-biometrics-data.mdx
Comment (leite08) [inline docs/devices-api/api-reference/get-biometrics-data.mdx:83]: nit: `;` is not a valid separator for JSON

File: docs/devices-api/more-info/webhooks.mdx
Comment (leite08) [inline docs/devices-api/more-info/webhooks.mdx:38]: If we were to follow the pattern used for `providers`, shouldn't we call this `devices`? Assuming that this field would also be used to report a device disconnection, `devices` would make more sense. But I don't know much about Tenovi's flow, so will leave it up to you, just a heads up.

File: packages/api/src/routes/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/routes/webhook/tenovi.ts:19]: How's auth being done? I don't see an auth lambda of sorts, and it seems its possible to have "Header-based authentication":

> Header-based authentication methods (e.g. Basic Auth) can be optionally included in the POST request, if required by the Client.

File: packages/api/src/shared/numbers.ts
Comment (leite08) [inline packages/api/src/shared/numbers.ts:7]: NABD, but why `.toFixed(2)`?

This is a good candidate for some unit testing! 🤓

Comment (RamilGaripov) [reply packages/api/src/shared/numbers.ts:7]: 2 decimal places seems like a reasonable cutoff for most measurements. Dima and I had this conversation a while ago when I was fixing a bug on weight measurements with Fitbit

File: packages/api/src/routes/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/routes/webhook/tenovi.ts:19]: This is something I was going to integrate next in a follow-up update. Could probably let the customer test the integration without auth?

File: docs/devices-api/more-info/webhooks.mdx
Comment (RamilGaripov) [reply docs/devices-api/more-info/webhooks.mdx:38]: Yea, it makes sense to keep the field names consistent! Good catch

File: packages/api/src/routes/middlewares/noauth.ts
Comment (leite08) [inline packages/api/src/routes/middlewares/noauth.ts:59]: Couldn't we always return an array, empty when nothing found?

File: packages/api/src/routes/middlewares/noauth.ts
Comment (leite08) [inline packages/api/src/routes/middlewares/noauth.ts:58]: Since this is supposed to have a comma-separated array, maybe add a TSDoc documenting this, as well as make the param to plural (`deviceIds`)?

File: packages/api/src/routes/middlewares/noauth.ts
Comment (leite08) [inline packages/api/src/routes/middlewares/noauth.ts:61]: no snake case in the middle of the code if possible, please

File: packages/api/src/providers/provider.ts
Comment (leite08) [inline packages/api/src/providers/provider.ts:20]: On DAPI, we call them `user`, leaving `patient` to MAPI. Can we change that here and everywhere this PR refers to patient?

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:81]: Trying to understand why this `else if`. Could it be another, separated `if`?

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:91]: nit: since this is related to the block from line 66 to 73, doesn't it make sense to have it right after that block? Just trying to make it easier to read.

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:73]: If I understand correctly, these update `biometrics`, right? I think it would make is clearer if we add that to the function names, something like `updateBiometricsWithBP(...)`.

Also, lets avoid `snake case` in the code, please. We just use it for DAPI payload responses b/c its heavy on units.

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:120]: nit: can we have the analytics after the `processRequest`? It seems better to first send the info to our cxs, then we do other stuff for us, like analytics.

File: packages/api/src/routes/webhook/tenovi.ts
Comment (leite08) [reply packages/api/src/routes/webhook/tenovi.ts:19]: Sounds good, let's ticket it so we don't forget it then - or add a checklist on the original ticket w/ one item specifically about this, please.

File: packages/api/src/shared/numbers.ts
Comment (leite08) [reply packages/api/src/shared/numbers.ts:7]: That's obtained by multiplying by 100 and converting to int. I don't see the reason for the `.toFixed(2)`. Did some [tests on JSFiddle](https://jsfiddle.net/2eya4ws3/5/), which confirms this.

NABD, but if it can be simplified its better for understanding.

Also, we have other places doing this, maybe we can refer to this new function?

File: packages/api/src/command/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/tenovi.ts:73]: Oh, right! Yea, working with biometrics units, which are in snake-case, kinda messes with me, and I start using it everywhere! ahaha already fixed it in a few places, but didn't stand out to me here at all! Good catch

File: packages/api/src/command/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/tenovi.ts:81]: Just one less matching operation. If the metric is `weight`, then we don't need to see if it's included in the `tenoviMetricTypes` cuz we already know that it is

File: packages/api/src/routes/middlewares/noauth.ts
Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/noauth.ts:59]: I made a decision to not do that, so that the payload doesn't have an empty array of `connectedDevices` if there's nothing to show. Same for the `devices` (previously, `newDevices`) array, if you're not connecting any new devices. I thought it looked nicer that way, but lmk if you agree or not

File: packages/api/src/routes/middlewares/noauth.ts
Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/noauth.ts:21]: Yea this makes sense

File: packages/api/src/providers/provider.ts
Comment (leite08) [reply packages/api/src/providers/provider.ts:20]: Taken to slack considering the nature of the discussion: https://metriport.slack.com/archives/C04DMKE9DME/p1692829416471799

File: packages/api/src/routes/middlewares/noauth.ts
Comment (leite08) [reply packages/api/src/routes/middlewares/noauth.ts:67]: Its fine if we want to massage the response/DTO, but IMO we should do that there. I think returning an array is cleaner: no elements has a clear representation on array, zero length.

File: packages/api/src/routes/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/routes/webhook/tenovi.ts:19]: Done!

File: packages/api/src/shared/numbers.ts
Comment (RamilGaripov) [reply packages/api/src/shared/numbers.ts:7]: Oh, yea, that makes sense

File: packages/api/src/routes/webhook/tenovi.ts
Comment (leite08) [reply packages/api/src/routes/webhook/tenovi.ts:19]: what's the link to the ticket? 😄

File: packages/api/src/routes/middlewares/noauth.ts
Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/noauth.ts]: Relics from another function

File: packages/api/src/providers/provider.ts
Comment (RamilGaripov) [reply packages/api/src/providers/provider.ts]: Solution: remove `patientId` and use `deviceUserId` instead, and not as an extraParam, but as an attribute in `ProviderMapItem`

File: packages/api/src/shared/numbers.ts
Comment (RamilGaripov) [reply packages/api/src/shared/numbers.ts]: Fixed

File: docs/devices-api/api-reference/get-biometrics-data.mdx
Comment (RamilGaripov) [reply docs/devices-api/api-reference/get-biometrics-data.mdx:29]: My Prettier makes these changes =( Had to disable it to revert this formatiing

File: packages/api/src/command/webhook/webhook.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/webhook.ts:35]: oh no, this is a relic from yesterday.. I thought I took out `patient_id` everywhere..

File: packages/api/src/routes/middlewares/connect-device.ts
Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/connect-device.ts:58]: Good catch!

Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/connect-device.ts:58]: I think the name of the function makes sense. I'll change the description to reflect the merging of connected devices into a list. But since it returns a list, I think `get...` makes more sense for the name

File: packages/api/src/command/connected-user/get-connected-user.ts
Comment (RamilGaripov) [reply packages/api/src/command/connected-user/get-connected-user.ts:133]: Not sure I understand what you mean. What index?

File: packages/api/src/routes/connect.ts
Comment (RamilGaripov) [reply packages/api/src/routes/connect.ts:191]: Added a return. Thank you!

File: packages/api/src/routes/middlewares/connect-device.ts
Comment (RamilGaripov) [reply packages/api/src/routes/middlewares/connect-device.ts:41]: Just following the Apple format here. I think N/A makes more sense.

File: packages/api/src/command/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/tenovi.ts:102]: Absolutely! Good catch

File: packages/api/src/command/connected-user/get-connected-user.ts
Comment (RamilGaripov) [reply packages/api/src/command/connected-user/get-connected-user.ts:133]: https://github.com/metriport/metriport-internal/issues/989

File: docs/devices-api/api-reference/connect/connect-provider.mdx
Comment (leite08) [reply docs/devices-api/api-reference/connect/connect-provider.mdx]: I see you moved one of the `Info`, but the other one on line 28 is still out of a `ParamField` - I'm assuming its intentional, right? Just to confirm.

Also, on line 29 we have a typo: "RMP" > "RPM"

@RamilGaripov

File: docs/devices-api/api-reference/get-biometrics-data.mdx
Comment (RamilGaripov) [reply docs/devices-api/api-reference/get-biometrics-data.mdx:31]: I'll create a ticket to change this across the docs. Also to change `metriportClient` to `metriport` and use multiline function calls to decrease horizontal scrolling

File: docs/devices-api/api-reference/get-biometrics-data.mdx
Comment (RamilGaripov) [reply docs/devices-api/api-reference/get-biometrics-data.mdx]: I have it enabled on-save

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:145]: Maybe add `data` to `extra` as well? Not sure if its too much data, though

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:112]: Left-over

**Review by leite08 — COMMENTED** [review]
Generally LGTM, just a few more comments, some defs NABD, but some need attention IMO.

File: packages/api/src/routes/connect.ts
Comment (RamilGaripov) [reply packages/api/src/routes/connect.ts]: I think this makes sense to me. I looked at the functions that are being called by this new `/connect/rpm/:provider` endpoint, I thought it necessary to add error handling to the `saveRpmDevice` function.

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:51]: no need to stringify, `capture.error` and `capture.message` already do this to stuff on `extra` 🤩

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:145]: ditto stringify extra

File: packages/api/src/command/webhook/tenovi.ts
Comment (leite08) [inline packages/api/src/command/webhook/tenovi.ts:38]: Curious, does it work - result in a string representation of the object? I'd think we need stringify for this since we don't implement `toString()` - maybe Zod does it out-of-the-box?

**Review by leite08 — APPROVED** [review]
LGTM!

Before we merge this, let's create a new upstream PR for the internal configs if we didn't yet and update this PR, please.

File: packages/api/src/command/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/tenovi.ts:38]: Ah, I think I misunderstood your previous comments on this! Fixed now

File: packages/api/src/command/webhook/tenovi.ts
Comment (RamilGaripov) [reply packages/api/src/command/webhook/tenovi.ts:51]: Oh, ill keep that in mind! Ty

**Pull request merged**
